File: 96_Unique_Binary_Search_Trees.py
from functools import lru_cache

class Solution:
    def numTrees(self, n: int) -> int:
        @lru_cache(None)
        def count_trees(nodes):
            if nodes <= 1:
                return 1
            
            total = 0
            for root in range(1, nodes + 1):
                left_trees = count_trees(root - 1)
                right_trees = count_trees(nodes - root)
                total += left_trees * right_trees
            
            return total
        
        return count_trees(n)

File: 96_Unique_Binary_Search_Trees.js
/**
 * @param {number} n
 * @return {number}
 */
var numTrees = function (n) {
  let memo = new Map();

  function countTrees(nodes) {
    if (nodes <= 1) {
      return 1;
    }
    if (memo.has(nodes)) {
      return memo.get(nodes);
    }

    let total = 0;
    for (let root = 1; root <= nodes; root++) {
      let leftTrees = countTrees(root - 1);
      let rightTrees = countTrees(nodes - root);
      total += leftTrees * rightTrees;
    }

    memo.set(nodes, total);
    return total;
  }

  return countTrees(n);
};

// Example usage:
console.log(numTrees(3)); // Output: 5
